lib/streamlit/runtime.py

```diff
@@ -0,0 +1,591 @@
+
+        Returns
+        -------
+        None
+
+        Threading
+        ---------
+        UNSAFE. Must be called on the eventloop thread.
+        """
+        await self._loop_coroutine(on_started)
+
+    def stop(self) -> None:
+        """Request that Streamlit close all sessions and stop running.
+        Note that Streamlit won't stop running immediately.
+
+        Threading
+        ---------
+        SAFE. May be called on any thread.
+        """
+        if self._state in (RuntimeState.STOPPING, RuntimeState.STOPPED):
+            return
+
+        LOGGER.debug("Runtime stopping...")
+        self._set_state(RuntimeState.STOPPING)
+        self._get_eventloop().call_soon_threadsafe(self._must_stop.set)
+
+    def is_active_session(self, session_id: str) -> bool:
+        """True if the session_id belongs to an active session.
+
+        Threading
+        ---------
+        SAFE. May be called on any thread.
+        """
+        # Dictionary membership is atomic in CPython, so this is thread-safe.
+        return session_id in self._session_info_by_id
+
+    def create_session(
+        self,
+        client: SessionClient,
+        user_info: Dict[str, Optional[str]],
+    ) -> str:
+        """Create a new session and return its unique ID.
+
+        Parameters
+        ----------
+        client
+            A concrete SessionClient implementation for communicating with
+            the session's client.
+        user_info
+            A dict that contains information about the session's user. For now,
+            it only (optionally) contains the user's email address.
+
+            {
+                "email": "example@example.com"
+            }
+
+        Returns
+        -------
+        str
+            The session's unique string ID.
+
+        Threading
+        ---------
+        UNSAFE. Must be called on the eventloop thread.
+        """
+        if self._state in (RuntimeState.STOPPING, RuntimeState.STOPPED):
+            raise RuntimeError(f"Can't create_session (state={self._state})")
+
+        session_data = SessionData(self._script_path, self._command_line or "")
+
+        session = AppSession(
+            event_loop=self._get_eventloop(),
+            session_data=session_data,
+            uploaded_file_manager=self._uploaded_file_mgr,
+            message_enqueued_callback=self._enqueued_some_message,
+            local_sources_watcher=LocalSourcesWatcher(session_data),
+            user_info=user_info,
+        )
+
+        LOGGER.debug(
+            "Created new session for client %s. Session ID: %s", id(client), session.id
+        )
+
+        assert (
+            session.id not in self._session_info_by_id
+        ), f"session.id '{session.id}' registered multiple times!"
+
+        self._session_info_by_id[session.id] = SessionInfo(client, session)
+        self._set_state(RuntimeState.ONE_OR_MORE_SESSIONS_CONNECTED)
+        self._has_connection.set()
```

Comment:
`_has_connection` used to be a `tornado.Condition`, but `asyncio.Condition` has different semantics (it requires a lock object, and doesn't work inside `asyncio.wait()` functions). Regardless, Event is more correct than Condition, because `_has_connection` has nothing to do with exclusive access to a resource: https://docs.python.org/3/library/asyncio-sync.html#condition